### DISCUSSION TITLE
OJS developer experience: live preview / hot reload

### DISCUSSION THREAD

**Discussion by declann** [quarto-dev/quarto-cli#11746] [Feature Requests]
### Description

Hi,

As a mainly JS developer using Quarto (and I love it!) I'm jealous of the developer experience available for Python/R in Quarto, Rstudio, and other tools supported by Posit - especially whenever iterating avoids full page reloads (I think this is a big part of why Notebooks and Rstudio are so popular).

In Quarto, the execution context for OJS is different versus Python/R: OJS runs in the browser, Python/R/Julia code runs in kernels (I think) that talk to Quarto during the render process. And OJS is functional and can add a tonne of value to outputs from Python/R or by itself.

But different execution contexts also means the developer experience for OJS is different.

Trying Rstudio, Positron, VSCode, I can't find any developer experience where it's possible to iterate on OJS in Quarto without full page reloads triggered by `quarto preview`. These are slow and break flow/context etc.

[Observable Framework](https://observablehq.com/framework/) shows that "[live previewing](https://observablehq.com/framework/getting-started#test-live-preview)" of OJS is possible, and that the OJS runtime (which Quarto uses) is suitable to support a super fast iterative developer experience for OJS.

Currently, I iterate in [Observable notebooks](https://observablehq.com/documentation/notebooks/) and migrate to Quarto when things are mostly ready - but this is really inefficient; and I especially run into trouble when my target is a Quarto page with significant different layout/features versus Observable notebooks (like a Quarto dashboard or a [closeread](https://closeread.dev)).

Observable nailed OJS DX in Framework and Observable notebooks, and I'm not under an illusion that achieving the same quality DX is a trivial task. `quarto preview` would need at least a socket server to communicate updates, and just identifying the changes to push will be a bit of work. And that doesn't account for non-OJS cells especially markdown/text updates, or how other particular environments should adapt (environment-specific changes is an alternative approach).

Am I missing a way to iterate on OJS without full page reloads?
Am I missing other big challenges to do this?
Do Quarto users even want this? I suspect most are Python and R developers and don't encounter OJS DX except briefly - I'm probably unusual because I make most of my work in OJS.
Are there public developments for OJS DX such as hot reload being considered/planned? (in `quarto preview` or in environments)

Imo, Quarto and OJS are great and for those developers who turn to OJS, a neat and fast DX can help them to do new and interesting things. Obviously there is a lot of people (like me) who develop in OJS outside of Quarto too.

Cheers,
Declan

**Comment by mcanouil** [comment]
Quarto needs to render the whole document to make a preview (possibly triggered by the watch input default setting).

Computations needs to occur even before rendering so there are no ways around that.

Observable or any raw HTML contents are different but still needs to be transformed into the AST internally to then be transformed as HTML output.

I don't believe what you want here is even possible for Quarto/Pandoc.

**Reply by declann** [reply]
Ah. Thanks for the response. So you would need to fall back to a full page refresh for other updates, possibly including markdown, possibly including new OJS cell positioning, which does seem kinda poor. At the same time, it could still tighten OJS iteration loop (I know it would help me even if limited - but I would probably really want hot markdown updates, possibly I'd consider moving my markdown into OJS blocks too in this scenario).

OTOH looking strictly at the OJS parts...

In my HTML on OJS heavy pages (like [this](https://calcwithdec.dev/posts/hearty-maths/)) I mainly see empty div containers.

![image](https://github.com/user-attachments/assets/8d6e4533-ff9e-4400-9efc-e4192eba6176)

The OJS runtime does a lot of heavy lifting already in the client side (necessarily).

Interesting basic requirements for OJS are already exposed. For instance [on that page](https://calcwithdec.dev/posts/hearty-maths/) you can paste `window._ojs.ojsConnector.mainModule.redefine("pinchiness_in", [], 0.4)` ([OJS runtime docs](https://github.com/observablehq/runtime?tab=readme-ov-file#moduleredefinename-inputs-definition)) into the console and OJS will alter the definition and reactively update (the slider will no longer work because the definition has changed)
[Screencast from 2024-12-28 05-37-02.webm](https://github.com/user-attachments/assets/6cf8c43a-3197-43ba-8523-0a57a8225158)

**Reply by mcanouil** [reply]
> Interesting basic requirements for OJS are already exposed. For instance on that page you can paste window._ojs.ojsConnector.mainModule.redefine("pinchiness_in", [], 0.4) (OJS runtime docs) into the console and OJS will alter the definition and reactively update (the slider will no longer work because the definition has changed)

Then you could be able to make a Quarto extension to allow live edition (note that it won't modify the source obviously).

**Reply by declann** [reply]
That seems like something it could be worthwhile I play with at some stage. Maybe I can begin by only doing hot reload for imported JS or imported OJS - that should help to know which changes to do a live reload or a full page reload for. The code in `quarto-ojs-runtime.js` seems to be well documented.

I think a blocker to deliver an extension however will be the preview server - the reload [socket message](https://github.com/quarto-dev/quarto-cli/blob/247513ca1c072567f44132de6aee3c474c5f5afc/src/core/http-devserver.ts#L187) is just "reload" when I update any file, so client code won't be able to conditionally reload without Quarto changes (extension spinning up a custom watcher is the only workaround I can think of)

**Comment by GuillaumeChretienCerema** [comment]
Hello everyone,
There’s been some news from Observable with the [Notebook Kit](https://observablehq.com/notebook-kit/kit), which is now based on Node.js and Vite. I’m sharing a small test (very limited to my specific use case) to make it easier to work with large pages in pure OJS (without R chunks). It’s just a little script that converts the {ojs} chunks from a .qmd file into an Observable-compatible .html page.
You can keep the Observable preview running in the background, work on your Quarto file, and simply run the conversion script—the web preview updates instantly! My dev skills are quite limited, and this is a proof of concept I put together in less than an hour, but it saves me from having to do everything on the Observable platform and then copy it back into Quarto.
Note: Debugging errors in this setup can be tricky, especially since the conversion process doesn't always provide clear error messages. Additionally, some formats like .xlsx are not yet supported by FileAttachment in this context.
I’ve created a demo GitHub repository (prerequisite: having Node.js installed): [qmd_2_ojs_live_preview](https://github.com/GuillaumeChretienCerema/qmd_2_ojs_live_preview) all is explained in the readme. It’s a bit of a hack, but who knows—maybe it’ll help someone!

**Comment by GuillaumeChretienCerema** [comment]
And to continue exploring the possibilities, if there’s a need to work chunk by chunk in real time, as on the platform: by opening the converted file my_file.observable.html in Visual Studio Code with the extension [https://github.com/GordonSmith/vscode-ojs](https://github.com/GordonSmith/vscode-ojs)
, it’s possible to run the chunks one by one and get the immediate result just below. (Be careful though: in that case you’ll need to manually copy the code back into the original Quarto file.)
Once again, this is a bit of a hack for a few use cases, but it makes it possible to stay 100% local for sensitive data and to get near–real-time renderings.

**Comment by mcanouil** [comment]
I did not look too much in details, but there is an extension for Observable.js (https://github.com/GordonSmith/vscode-ojs).

As many features come from third party extensions (R and Python notably), it might be possible to do the same (https://github.com/quarto-dev/quarto)
